File: intern/1498.number-of-subsequences-that-satisfy-the-given-sum-condition.js
/*
 * @lc app=leetcode id=1498 lang=javascript
 *
 * [1498] Number of Subsequences That Satisfy the Given Sum Condition
 */

// @lc code=start
const MOD = 10 ** 9 + 7;

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var numSubseq = function (nums, target) {
    nums.sort((a, b) => a - b);
    let count = 0;

    let left = 0;
    let right = nums.length - 1;

    while (left <= right) {
        if (nums[left] + nums[right] > target) {
            right--;
        } else {
            // 对于 nums[left+1:right]，有多少种子序列，前面加上 nums[left] 就是所有子序列
            // 不会发生重复，因为 left 不会被重复取到
            count += mod2Pow(right - left);
            count %= MOD;
            left++;
        }
    }

    return count;
};

const cache = new Map();
function mod2Pow(exp) {
    if (exp === 0) {
        return 1;
    }

    if (cache.has(exp)) {
        return cache.get(exp);
    }
    const result = (mod2Pow(exp - 1) * 2) % MOD;
    cache.set(exp, result);
    return result;
}
// @lc code=end

File: src/2025/1498.number-of-subsequences-that-satisfy-the-given-sum-condition.cpp
/*
 * @lc app=leetcode id=1498 lang=cpp
 *
 * [1498] Number of Subsequences That Satisfy the Given Sum Condition
 */

#include <algorithm>
#include <cinttypes>
#include <vector>

// @lc code=start
class Solution {
 public:
  int numSubseq(std::vector<int>& nums, const int target) {
    std::ranges::sort(nums);
    std::intmax_t result = 0;
    for (size_t left = 0; left < nums.size(); left++) {
      const int right_target = target - nums[left];
      const size_t right = BinarySearchUpperBound(nums, right_target);
      if (right <= left) {
        break;
      }
      const size_t between_element_count = right - left - 1;
      result += PowMod(2, between_element_count, kMod);
      result %= kMod;
    }
    return static_cast<int>(result);
  }

 private:
  static constexpr int kMod = 1e9 + 7;

  static size_t BinarySearchUpperBound(const std::vector<int>& nums,
                                       const int target) {
    size_t left = 0;
    size_t right = nums.size();

    while (left < right) {
      const size_t mid = (right - left) / 2 + left;
      if (nums[mid] <= target) {
        left = mid + 1;
      } else if (nums[mid] > target) {
        right = mid;
      }
    }

    return left;
  }

  static std::intmax_t PowMod(const std::intmax_t base,
                              const size_t exp,
                              const std::intmax_t mod) {
    std::intmax_t result = 1;
    size_t left_exp = exp;
    std::intmax_t current_base = base;

    while (left_exp > 0) {
      if (left_exp % 2) {
        result *= current_base;
        result %= mod;
      }
      current_base *= current_base;
      current_base %= mod;
      left_exp /= 2;
    }

    return result;
  }
};
// @lc code=end

int main() {
  Solution sol;
  std::vector<int> nums = {3, 5, 6, 7};
  sol.numSubseq(nums, 9);
}
